packages/fields/src/types/Content/slate-walker.js

```diff
@@ -0,0 +1,64 @@
+import { identity } from '@keystone-alpha/utils';
+const noop = () => {};
+
+// A depth-first, top-down tree walking algorithm.
+function visitNode(node, visitors) {
+  const recurse = childNode => visitNode(childNode, visitors);
+  let visitedNode = null;
+
+  // Registered visitors might serialise a node.
+  // If they do, it's their responsibility to also serialise all the child
+  // nodes.
+  // If they don't, they return falsey and we pass it to the default visitor to
+  // handle.
+  switch (node.object) {
+    case 'document': {
+      visitedNode = visitors.visitDocument(node, recurse);
+      break;
+    }
+    case 'block': {
+      visitedNode = visitors.visitBlock(node, recurse);
+      break;
+    }
+    case 'inline': {
+      visitedNode = visitors.visitInline(node, recurse);
+      break;
+    }
+    case 'text': {
+      visitedNode = visitors.visitText(node, recurse);
+      break;
+    }
+    default: {
+      throw new Error(`Encountered unknown type '${node.object}' in Slate document`);
```

Comment:
I think `mark` is another type that should be here as well.

Reply:
`mark`s only show up as children of `leaves`. This algorithm only walks `nodes`.

Here's an example for the text

> H*el*lo

```json
{
  "object": "document",
  "data": {},
  "nodes": [
    {
      "object": "block",
      "type": "paragraph",
      "data": {},
      "nodes": [
        {
          "object": "text",
          "leaves": [
            {
              "object": "leaf",
              "text": "H",
              "marks": []
            },
            {
              "object": "leaf",
              "text": "el",
              "marks": [
                {
                  "object": "mark",
                  "type": "bold",
                  "data": {}
                }
              ]
            },
            {
              "object": "leaf",
              "text": "lo",
              "marks": []
            }
          ]
        }
      ]
    }
  ]
}
```